### DISCUSSION TITLE
How to dynamically load icons?

### DISCUSSION THREAD

**Discussion by twilio-jyoung** [twilio-labs/paste#2765] [Q&A] [answered]
I am building content in a CMS, and need to be able to allow users to pick the Icons they would like to use.  

Can you add a code sample to the docs that just takes the icon name as a parameter and lazy loads the icon?

I was able to get it working, but with boatloads of warnings.

[localhost-1666634838384.log](https://github.com/twilio-labs/paste/files/9854172/localhost-1666634838384.log)

![Screen Shot 2022-10-24 at 1 08 32 PM](https://user-images.githubusercontent.com/44909531/197595557-7f52b39d-a64d-4757-ad50-17b7d0eb293c.png)

paste-icon-loader.js
```
import { lazy, Suspense } from "react";

const iconData = require("@twilio-paste/icons/json/icons.json");

export default async function getIcon(name, iconProps) {
	let iconName = `${name}Icon`;

	// check to make sure the icon name is valid
	if (typeof iconData.icons.find((x) => x.name == iconName) === undefined) {
		console.log(
			`unable to find ${iconName} in the list of available icons from paste`
		);
		return null;
	}

	// lazy load the icon
	let PasteIcon = lazy(() =>
		import(
			/* webpackMode: "eager" */ `@twilio-paste/icons/esm/${iconName}`
		).then((module) => ({
			default: module[iconName],
		}))
	);

	// set required props if none passed in
	if (!iconProps) iconProps = {};
	if (!iconProps.decorative) iconProps.decorative = true;
	if (!iconProps.title) iconProps.title = "default title";

	// build suspense wrapper around lazy load component
	let result = (
		<Suspense fallback={<div>Loading...</div>} key={iconName}>
			<PasteIcon {...iconProps} key={iconName} />
		</Suspense>
	);

	return result;
}

```

another-file.js
```
		let iconComponent = await getIcon("Accept", {
			decorative: false,
			title: "Accept Icon",
			size: "sizeIcon30",
		});
		
		flex.MainHeader.Content.add(iconComponent);
```

**Comment by gloriliale** [comment, marked as answer]
Hi @twilio-jyoung! It looks like twilio-flex.min.js is loading in d.ts files. Is there a reason those are being bundled into your build and loaded into the browser?

**Reply by twilio-jyoung** [reply]
i dont think that flex is the issue here.  twilio-flex.min.js is just what everything gets condensed down into when building plugins.  

to ensure that wasnt the issue, i ported this code to a net new create-react-app, same errors.  
![Screen Shot 2022-10-24 at 10 31 37 PM](https://user-images.githubusercontent.com/44909531/197676009-4a1cb613-9ec0-4519-bdcc-8b64b7017a67.png)

[errors.txt](https://github.com/twilio-labs/paste/files/9856914/errors.txt)

**Reply by TheSisb** [reply]
Hi @twilio-jyoung,

The error seems to be around [`*.d.ts` files](https://en.wikipedia.org/wiki/TypeScript#Declaration_files):
![image](https://user-images.githubusercontent.com/1592327/197818056-de1c09aa-a6e3-455b-bfa8-9c76c1d96236.png)

These files should not be loaded by the browser, they're used by Typescript for typechecking at dev and build time. Those errors seem to be warning at run-time, which is expected if those files somehow get bundled into `twilio-flex.min.js`.

I don't think this is something we can fix from our end; no one else has reporting this so far. My hunch is this is a configuration issue with your build pipeline.

**Reply by twilio-jyoung** [reply]
can you provide a code sample in the docs showing what I am attempting?  i need to dynamically load the icon at runtime given only the icon name.

**Reply by TheSisb** [reply]
Here's an example of how they're doing it in Console

```js
import { Spinner, SpinnerProps } from "@twilio-paste/core/spinner";
import { ErrorIcon } from "@twilio-paste/icons/esm/ErrorIcon";
import { IconWrapperProps } from "@twilio-paste/icons/esm/helpers/IconWrapper";
import React, { Component, ReactNode, Suspense } from "react";
import { useTranslation } from "react-i18next";

import { safeLazyLoad } from "../../shared/helpers/safeLazyLoad";

interface IconErrorBoundaryProps {
  children: ReactNode;
}
interface IconErrorBoundaryState {
  hasError: boolean;
}

class IconErrorBoundary extends Component<
  IconErrorBoundaryProps,
  IconErrorBoundaryState
> {
  constructor(props: IconErrorBoundaryProps) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError(): IconErrorBoundaryState {
    return { hasError: true };
  }

  render(): JSX.Element | ReactNode {
    if (this.state.hasError) {
      // NOTE: not sure if showing another icon here
      //       is enough signal to the developer that
      //       something went wrong. Though there is
      //       a wall of error text on the console.
      return (
        <ErrorIcon
          title="an error occurred"
          color="colorTextError"
          decorative
        />
      );
    }

    return this.props.children;
  }
}

type IconLoaderProps = Omit<SpinnerProps, "title"> &
  IconWrapperProps & {
    icon: string;
    title?: string;
    decorative: boolean;
  };
const IconLoader: React.FC<IconLoaderProps> = React.memo(
  ({ icon, ...iconProps }: IconLoaderProps) => {
    const AsyncIcon = safeLazyLoad(
      import(`@twilio-paste/icons/esm/${icon}.js`).then((c) => ({
        default: c[icon],
      }))
    );
    const { t } = useTranslation("common");
    return (
      <IconErrorBoundary>
        <Suspense
          fallback={
            <Spinner
              size="sizeIcon30"
              {...iconProps}
              title={t("Loading")}
              decorative
            />
          }
        >
          <AsyncIcon icon={icon} {...iconProps} />
        </Suspense>
      </IconErrorBoundary>
    );
  }
);
export { IconLoader };
```

**Reply by twilio-jyoung** [reply]
this looks like what I need, would you mind pointing me to safeLazyLoad?

**Reply by TheSisb** [reply]
Sent on Slack 👍  Glad to be able to help!
